draft/0002-subdocapi.md

```diff
@@ -0,0 +1,351 @@
+
+Subdocument operations come in two flavors. One is the simple single-operation
+form which takes a single document ID and a single path. Another is a
+_multi-path_ form which retrieves or mutates _several_ paths within a document.
+
+Single-path methods are slightly more efficient and require less processing
+overhead than their multi-path equivalents (that is, even if you pass a single
+path to a multipath method, some additional processing overhead may be involved).
+
+Note that both 'multi-path' and 'single-path' APIs operate on the same
+document, and correspond to a single binary operation. This should not be
+confused with 'multi' operations in some SDKs which pipeline multiple binary
+requests in the network.
+
+## Objects
+
+MetaDocFragment
+
+This subclass contains the same fields as the MetaDoc, except the `value`
+may wish to be replaced by the `fragment` field. This is to avoid confusion
+regarding whether the value "truly" contains the entire document value, or
+merely a subdoc.
+
+The `MetaDoc` object itself is the return value from normal KV operations
+which contain the value, cas, mutation token, etc.
+
+This object is used as an example, some SDKs may wish to either use the
+'value' or 'document' field of the existing `MetaDoc` return value, or
+subclass it, replacing value with 'fragment'.
+
+## Path
+
+From the perspective of the SDK, the path is a simple string used to specify
+a location within the document. The path uses N1QL syntax to specify locations
+within a document, thus a valid path may be:
+
+    foo.bar.baz
+
+or
+
+    `foo`.`bar`.`baz`
+
+or
+
+    `foo.some_array[0]`
+
+or
+
+    `foo.some_array[-1]`
+
+As in N1QL, special path characters may be escaped via a backtick (`` ` ``).
+
+Some operations (notably, those taking a path to an array) may accept the
+_empty path_ which refers to the entire document. This is valid if the document
+is a bare JSON array.
+
+Each path component must be valid JSON (without the enclosing quotes). This
+means that things like tabs, quotes, and backslashes must be escaped. To think
+of it another way, the path matching algorithm in the server works via simple
+string comparison, so a JSON key `"bs\\ and \t tabs are here" would be matched
+by a path component of `"bs\\ and \t tabs are here"`.
+
+The SDK MUST verify that a path is non-empty for operations which do not
+support empty paths. All other forms of path verification is done on the server
+(but an empty path for unsupported operations will return a memcached `EINVAL`
+which isn't really helpful).
+
+## Single-path Methods
+
+Methods are named exactly like their top-level counterparts, and should also
+exist at the top level as well.
+
+Error codes are common to most operations.
+
+### `get_in(docid, path)`
+Gets a single path from the document. Returns a MetaDocFragment containing the
+path value.
+
+* `docid`: Document ID
+* `path`: The sub-document path
+
+### `exists_in(docid, path)`
+
```

Comment:
That’s news to me!

If this is the case then it should be idiomatic to the SDK. I consider exists a “valueless get”; in any event I should probably change the description to indicate that exists is successful by lack of an error
> On Dec 30, 2015, at 1:36 AM, Michael Nitschinger <notifications@github.com> wrote:
> 
> In draft/0002-subdocapi.md <https://github.com/couchbaselabs/sdk-rfcs/pull/2#discussion_r48594383>:
> 
> > +## Single-path Methods
> > +
> > +Methods are named exactly like their top-level counterparts, and should also
> > +exist at the top level as well.
> > +
> > +Error codes are common to most operations.
> > +
> > +### `get_in(docid, path)`
> > +Gets a single path from the document. Returns a MetaDocFragment containing the
> > +path value.
> > +
> > +* `docid`: Document ID
> > +* `path`: The sub-document path
> > +
> > +### `exists_in(docid, path)`
> > +
> I'm asking because we already have bucket-level exists method which returns a boolean (based on observe).
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/couchbaselabs/sdk-rfcs/pull/2/files#r48594383>.
>